File: linked_list/reverse_linked_list.js
// SOLVED

// Input: 1->2->3->4->5->NULL
// Output: 5->4->3->2->1->NULL

class LinkedListNode {
  constructor(val) {
    this.val = val;
    this.next = null;
  }
}

let a = new LinkedListNode(1);
let b = new LinkedListNode(2);
let c = new LinkedListNode(3);
let d = new LinkedListNode(4);
let e = new LinkedListNode(5);

a.next = b;
b.next = c;
c.next = d;
d.next = e;

var reverseList = function(head) {
  if (head === null) {
    return head
  }
  let node = head;
  let previous, next;
  while (node) {
    // save node.next before overwriting it
    next = node.next;
    // overwrite node.next with previous node
    node.next = previous;
    // step forward in the list
    previous = node;
    node = next;
  }
  return previous;
}

reverseList(a)

File: Ruby/linked_list/reverse_linked_list.rb
# https://leetcode.com/explore/learn/card/linked-list/219/classic-problems/1205/

# Reverse a singly linked list.

# Input: 1->2->3->4->5->NULL
# Output: 5->4->3->2->1->NULL

# A linked list can be reversed either iteratively or recursively. Could you implement both?

# Definition for singly-linked list.
class ListNode
    attr_accessor :val, :next
    def initialize(val)
        @val = val
        @next = nil
    end
end

def return_list_values(head)
    arr = []
    curr = head
    while curr.next != nil
        arr << curr.val
        curr = curr.next
    end
    arr << curr.val
end

# using the orginal head node, move every next node to the front
# save the newHead (node.next)
# overwrite it (node.next = node.next.next)
# now move the newHead to the beginning (newHead.next = head)
# and finish the loop by resetting the head (head = newHead)
# runtime O(n), space O(1)
def reverse_list(head)
    return nil if !head

    node = head

    # node is always pointing at the original head
    while node.next
        newHead = node.next
        node.next = node.next.next
        newHead.next = head
        head = newHead
    end

    head
end

# could also solve this problem with recursion
# runtime & space would be O(n)
